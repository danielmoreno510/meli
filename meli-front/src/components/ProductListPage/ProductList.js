import { useHistory } from "react-router-dom";
import PropTypes from "prop-types";

import { formatCurrency } from "../../utils/currencyUtils";

const ProductList = ({ productList, getProduct }) => {
  const history = useHistory();
  const goToProduct = (product) => () => {
    history.push(`/items/${product.id}`);
    getProduct(product.id);
  };

  return productList ? (
    productList.map((product) => (
      <div
        key={product.id}
        className="result-product"
        data-testid="result-product"
      >
        <div className="reuslt-image-container">
          <img
            className="result-image"
            src={product.picture}
            alt={product.title}
            onClick={goToProduct(product)}
            data-testid="result-image"
          />
        </div>
        <div className="container-result">
          <div className="content-result">
            <div className="result-amount">
              {formatCurrency(product.price.currency, product.price.amount)}
            </div>
            <div className="result-city">{product.city_name}</div>
          </div>
          <h1 className="result-title">{product.title}</h1>
          <div className="result-condition">{product.condition}</div>
        </div>
      </div>
    ))
  ) : (
    <div />
  );
};

ProductList.propTypes = {
  productList: PropTypes.array,
  getProduct: PropTypes.func,
};

export default ProductList;
