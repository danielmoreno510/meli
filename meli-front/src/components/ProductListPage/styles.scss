.result-product {
  display: flex;
  padding-top: 20px;
  padding-bottom: 20px;
  border-bottom: 1px solid #dddddd;
}

.result-image {
  width: 140px;
  margin-right: 24px;
  cursor: pointer;
}

.container-result {
  flex: 1;
}

.result-title {
  font-size: 22px;
  font-weight: 300;
  margin: 0px !important;
}

.content-result {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.result-amount {
  font-size: 24px;
  font-weight: 500;
  margin-top: 14px;
}

.result-condition {
  margin-top: 20px;
}
