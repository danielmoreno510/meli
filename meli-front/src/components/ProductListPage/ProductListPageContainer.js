import ProductListPage from "./ProductListPage";
import { fetchProduct, fetchProductList } from "../../state/actions";
import { StoreService } from "../../state/StoreService.js";

export const mapStateToProps = (state) => ({
  productList: state.products.productList,
  isFetchingProductList: state.products.isFetchingProductList,
  hasSearch: state.products.hasSearch
});

export const mapDispatchToProps = (dispatch) => {
  return {
    getProductList: (search) => {
      dispatch(fetchProductList(search));
    },
    getProduct: (id) => {
      dispatch(fetchProduct(id));
    }
  };
};

const ProductListPageContainer = StoreService.connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductListPage);
export default ProductListPageContainer;
