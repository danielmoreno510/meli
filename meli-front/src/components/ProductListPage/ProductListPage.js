import React, {useEffect} from "react";
import { useHistory, useLocation } from "react-router-dom";
import queryString from "query-string";
import PropTypes from "prop-types";

import { formatCurrency } from "../../utils/currencyUtils";
import WithLoadingIndicator from "../WithLoading";
import "./styles.scss";

export const ProductList = ({ productList, getProduct }) => {
  const history = useHistory();

  return productList ? (
    productList.map((product) => (
      <div
        key={product.id}
        className="result-product"
        data-testid="result-product"
      >
        <div>
          <img
            className="result-image"
            src={product.picture}
            alt={product.title}
            onClick={() => {
              history.push(`/items/${product.id}`);
              getProduct(product.id);
            }}
            data-testid="result-image"
          />
        </div>
        <div className="container-result">
          <h1 className="result-title">{product.title}</h1>
          <div className="content-result">
            <div className="result-amount">
              {formatCurrency(product.price.currency, product.price.amount)}
            </div>
            <div>{product.city_name}</div>
          </div>
          <div className="result-condition">{product.condition}</div>
        </div>
      </div>
    ))
  ) : (
    <div></div>
  );
};
const ProductsWithLoadingIndicator = WithLoadingIndicator(ProductList);

const ProductListPage = ({
  hasSearch,
  productList,
  isFetchingProductList,
  getProductList,
  getProduct,
}) => {
  const location = useLocation();
  const value = queryString.parse(location.search);

  useEffect(() => {
    !hasSearch && getProductList(value.search);
  }, [value.search]);

  return (
    <ProductsWithLoadingIndicator
      productList={productList}
      isFetching={isFetchingProductList}
      getProduct={getProduct}
    />
  );
};

ProductListPage.propTypes = {
  productList: PropTypes.array,
};

export default ProductListPage;
