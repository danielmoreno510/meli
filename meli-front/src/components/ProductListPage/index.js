import ProductListPageContainer from './ProductListPageContainer'

export default ProductListPageContainer