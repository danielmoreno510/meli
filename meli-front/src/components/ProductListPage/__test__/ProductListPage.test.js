import React from "react";
import { shallow } from "enzyme";

import ProductListPage from "../ProductListPage";

jest.mock("react-router-dom", () => {
  return {
    useLocation: jest
      .fn()
      .mockReturnValue({ pathname: "/items", search: "?search=iphone%2012" }),
  };
});

describe("ProductListPage component", () => {
  const testProps = {
    isFetchingProductList: false,
    productList: [
      {
        id: "MLA916044386",
        title: "iPhone 11 64 Gb Blanco",
        installments: { currency_id: "ARS" },
        price: { currency: "ARS", amount: "123456" },
        address: { city_name: "Mar de la plata" },
      },
    ],
    getProduct: jest.fn()
  };

  describe("WHEN the ProductListPage component is render", () => {
    const wrapper = shallow(<ProductListPage {...testProps} />);

    it("THEN should display ProductListPage component", () => {
      expect(wrapper).toMatchSnapshot();
    });

    const productList = wrapper
      .find("ComponentWithLoadingIndicator")
      .dive()
      .find("ProductList");

    it("THEN should exist the productList component", () => {
      expect(productList.exists()).toBeTruthy();
    });
  });
});
