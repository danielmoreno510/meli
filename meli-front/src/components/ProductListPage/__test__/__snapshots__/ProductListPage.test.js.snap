// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListPage component WHEN the ProductList component is render THEN should display ProductList component 1`] = `
<div
  className="result-product"
  data-testid="result-product"
  key="MLA916044386"
>
  <div>
    <img
      alt="iPhone 11 64 Gb Blanco"
      className="result-image"
      data-testid="result-image"
      onClick={[Function]}
    />
  </div>
  <div
    className="container-result"
  >
    <h1
      className="result-title"
    >
      iPhone 11 64 Gb Blanco
    </h1>
    <div
      className="content-result"
    >
      <div
        className="result-amount"
      >
        ARSÂ 123,456.00
      </div>
      <div />
    </div>
    <div
      className="result-condition"
    />
  </div>
</div>
`;

exports[`ProductListPage component WHEN the ProductList component is render without props THEN should display ProductList component 1`] = `<div />`;

exports[`ProductListPage component WHEN the ProductListPage component is render THEN should display ProductListPage component 1`] = `
<ComponentWithLoadingIndicator
  getProduct={[MockFunction]}
  isFetching={false}
  productList={
    Array [
      Object {
        "address": Object {
          "city_name": "Mar de la plata",
        },
        "id": "MLA916044386",
        "installments": Object {
          "currency_id": "ARS",
        },
        "price": Object {
          "amount": "123456",
          "currency": "ARS",
        },
        "title": "iPhone 11 64 Gb Blanco",
      },
    ]
  }
/>
`;
